### ISSUE TITLE
Busy loop hang in PiscesRenderingEngine in OpenJDK 8

### ISSUE THREAD

**Issue by Matze2** [AdoptOpenJDK/openjdk-build#716]
As discussed in the Slack #general...

# Problem Description
While testing our application with AdoptOpenJDK instead of Oracle JDK we faced a busy loop problem that is maybe similar to the one described for Marlin (Java 9+) in https://bugs.java.com/bugdatabase/view_bug.do?bug_id=8191815

In our case sun.java2d.pisces.Dasher.lineTo(NaN, NaN) is called. But it seems that also using 0 and very high values as described in the OpenJDK Bug above also causing an endless loop.
The "NaN" problem seems to be in Pisces only since Oracle JDK and Zulu do not have this problem. Oracle uses the closed source Ductus engine in Java 8, Zulu seems to use a backport of Marlin. Both engines seem to catch "senseless" values in advance.

In our case this scenario is triggered in a web application by JFreeChart. In the end all threads would hang in busy loops which could lead render an host completely unusable.

# How to reproduce
* Compile and run the attached class
* It will print "If you don't see 'Not hanging.' then you hang :)" in console
* If the bug is not triggered you will see a second line "Not hanging."
* If the bug is triggered it will stay in line 29 in g2.draw(line) forever and consuming 100% CPU

[EndlessLoop.java.txt](https://github.com/AdoptOpenJDK/openjdk-build/files/2553559/EndlessLoop.java.txt)

# Tested distributions (all on Windows 10)
* AdoptOpenJDK jdk8u172-b11 triggers the problem
* AdoptOpenJDK jdk8u181-b13 triggers the problem
* AdoptOpenJDK jdk8u192-b12 triggers the problem
* AdoptOpenJDK jdk8u192-b13-0.11.0 triggers the problem (so also with OpenJ9)
* zulu8.31.0.1-jdk8.0.181-win_x64 does not trigger the problem (uses a backport of MarlinRenderingEngine)
* Oracle jre1.8.0_181 does not trigger the problem (also uses DuctusRenderingEngine)

So the problem seems to be related to the implementation in Pisces.

If you need more information just let me know.

**Comment by Matze2** [comment]
A valid workaround seems to be the use of the Marlin renderer from https://github.com/bourgesl/marlin-renderer

I did the following steps:
- Downloaded marlin jar from the github page above
- Added the following VM arguments to the command line:
    `-Dsun.java2d.renderer=org.marlin.pisces.MarlinRenderingEngine -Xbootclasspath/a:/path/to/marlin-0.9.3-Unsafe.jar`

This cured the described problem.

**Comment by karianna** [comment]
I'm going to reopen this as a docs issue for our support site.  @Matze2 Thanks for your excellent analysis.

**Comment by Matze2** [comment]
I was asked to summarize all our findings about visual and functional differences between Oracle JDK and AdoptOpenJDK (and Zulu) regarding Graphics2D and their possible workarounds.

## Busy loop hang in PiscesRenderingEngine 
This refers to this issue. The workaround is to replace Pisces with Marlin using the following steps:

* Download marlin jar from https://github.com/bourgesl/marlin-renderer/releases (I used marlin-0.9.3-Unsafe.jar)
* Added the following VM arguments to the command line:
    -Dsun.java2d.renderer=org.marlin.pisces.DMarlinRenderingEngine -Xbootclasspath/a:/path/to/marlin-0.9.3-Unsafe.jar

## Arbitrary long hanging in Marlin when drawing an erratic long line (possible DoS)
Related Marlin Issue: https://github.com/bourgesl/marlin-renderer/issues/31

Solution is to add the following rendering hint to Marlin:
```
Graphics2D g2 = (Graphics2D)g;
g2.setRenderingHint(RenderingHints.KEY_ANTIALIASING, RenderingHints.VALUE_ANTIALIAS_ON);
```

## Condensed 90° rotated texts drawn in Graphics2d

Basically generated by the following code:
```
Graphics2D g2 = (Graphics2D)g;
Font font = g2.getFont();
AffineTransform affineTransform = new AffineTransform();
affineTransform.rotate(Math.toRadians(-90), 0, 0);
Font rotatedFont = font.deriveFont(affineTransform);
g2.setFont(rotatedFont);
g2.drawString(text, x, y);
```
Using AdoptOpenJDK the text appears very condensed compared to Oracle and also Zulu.
On Linux this seems to be worse than on Windows which might be related to different fonts but the effect is visible in both cases.

No real workaround here but an observation:
Replacing fontmanager.dll and freetype.dll with the ones from Zulu fixes the problem.

**Comment by wmarkow** [comment]
@Matze2 
>In our case this scenario is triggered in a web application by JFreeChart. In the end all threads would hang in busy loops which could lead render an host completely unusable.

In my case it was also triggered by **JFreeChart-1.0.14**. It was already reported and fixed in [this issue](https://sourceforge.net/p/jfreechart/bugs/1121/).

**Comment by bourgesl** [comment]
See #1119

**Comment by M-Davies** [comment]
@karianna Can this be closed? It looks like discussions for this are best continued in #1119

**Comment by karianna** [comment]
> @karianna Can this be closed? It looks like discussions for this are best continued in #1119

Yes, effectively a duplicate
